src/librustc/ty/layout.rs

```diff
@@ -919,240 +919,242 @@ impl<'a, 'gcx, 'tcx> Layout {
             }
 
             // ADTs.
-            ty::TyStruct(def, substs) => {
-                if ty.is_simd() {
-                    // SIMD vector types.
-                    let element = ty.simd_type(tcx);
-                    match *element.layout(infcx)? {
-                        Scalar { value, .. } => {
-                            return success(Vector {
-                                element: value,
-                                count: ty.simd_size(tcx) as u64
-                            });
-                        }
-                        _ => {
-                            tcx.sess.fatal(&format!("monomorphising SIMD type `{}` with \
-                                                     a non-machine element type `{}`",
-                                                    ty, element));
+            ty::TyAdt(def, substs) => match def.adt_kind() {
+                AdtKind::Struct => {
```

Comment:
I don't think this is a good idea, because the old `TyStruct` is similar to the one-variant `TyEnum` case.
It would probably be easier to have `ty.is_simd()` (can that be done on the `TyAdt` `def` instead?) as a guard, and then combine the rest into the previous `TyEnum` handling.